### PR TITLE
Use constant address space for SPIR-V parameters when generating tessellation control shader.

### PR DESCRIPTION
**Pull request by mbarriault**
In MoltenVK, `MVKCommandEncoder::setComputeBytes` can call either `setBytes` or `setBuffer`, but Metal only allows `setBytes` to be used with the `constant` address space. As a small read-only buffer, this should be a safe change for both cases.

### CODE REVIEW

**Comment by cdavis5e** [comment]
Only one problem: The `constant` address space imposes an alignment requirement of 256 bytes. For an indirect draw, this buffer comes from the indirect draw parameters, which we can't guarantee will be 256-byte aligned. I think this needs to be fixed MoltenVK-side.

**Comment by mbarriault** [comment]
I have a number of other fixes related to tessellation in MoltenVK and a few bug reports for issues I couldn't solve, I'll file that bug report there too and leave this open for now.

**Comment by mbarriault** [comment]
@cdavis5e Do you have documentation on this 256-byte requirement? I can't find it in the MSL spec nor MTLComputeCommandEncoder documentation.

**Comment by cdavis5e** [comment]
> @cdavis5e Do you have documentation on this 256-byte requirement? I can't find it in the MSL spec nor MTLComputeCommandEncoder documentation.

From `-[MTLComputeCommandEncoder setBuffer:atOffset:]`:

> For buffers in the constant address space, the offset must be aligned to 256 bytes in macOS.

So iOS does not have this restriction, but macOS does.

**Comment by mbarriault** [comment]
Okay, noted, thanks!

**Comment by HansKristian-Work** [comment]
My primary concern here is alignment indeed. I'll let @cdavis5e have the final word on this PR.

**Comment by mbarriault** [comment]
The pull request has been updated to fix the tests for this change.

File: spvpatch.patch [hunk 1]
```diff
@@ -0,0 +1,39 @@
+diff --git a/spirv_msl.cpp b/spirv_msl.cpp
```
Comment (HansKristian-Work) [inline spvpatch.patch]: ?

Comment (mbarriault) [reply spvpatch.patch]: Whoops, that's another change I've been toying with but don't have the full understanding yet to commit. Shouldn't have been included, not sure how it got caught up, will update pull request shortly to remove it.

**Comment by cdavis5e** [comment]
So, I've thought about this, and I think what we should do is: the address space should be changed to `constant`, but *only* for tessellation control shaders. MoltenVK always uses `setComputeBytes()` here, even in indirect draws, so we don't have the issue with buffer alignment here.

But for vertex shaders, we need to keep using the `device` AS.

**Comment by mbarriault** [comment]
This change is indeed only for tessellation control shaders. However MoltenVK doesn't strictly use `setComputeBytes`: if dynamic Metal buffer feature isn't enabled, it falls back on `setComputeBuffer`.

**Comment by cdavis5e** [comment]
> This change is indeed only for tessellation control shaders.

Are you sure? The changes to the tests indicate that it also affects vertex shaders.

> However MoltenVK doesn't strictly use `setComputeBytes`: if dynamic Metal buffer feature isn't enabled, it falls back on `setComputeBuffer`.

Yes, but in that case it gets a temporary buffer and ensures that the buffer offset is aligned correctly for the `constant` AS.

**Comment by mbarriault** [comment]
Yes, you're right. In `CompilerMSL::entry_point_args_builtin` and `MVKCmdDrawIndirect::encode`. What are the performance characteristics of `device` vs `constant`? My intuition (OpenCL background) suggests that, for these buffers, `constant` should be preferred, but as an array of `uint32_t`s that would be rather wasteful in memory.

I'll update to your suggestion and only use `constant` in the tessellation control case.

**Comment by mbarriault** [comment]
Hm, nope, that won't do either. The non-indirect case still uses `setVertexBytes` and crashes with `device`.

**Comment by cdavis5e** [comment]
KhronosGroup/MoltenVK#572 should fix that.

**Comment by mbarriault** [comment]
Updated pull request for merged KhronosGroup/MoltenVK#572.

File: spirv_msl.cpp [hunk 2]
```diff
@@ -5608,13 +5608,20 @@ void CompilerMSL::entry_point_args_builtin(string &ep_args)
...
+		if (get_execution_model() == ExecutionModelTessellationControl)
 		{
 			if (!ep_args.empty())
 				ep_args += ", ";
 			ep_args +=
-			    join("device uint* spvIndirectParams [[buffer(", msl_options.indirect_params_buffer_index, ")]]");
-		}
+			    join("constant uint* spvIndirectParams [[buffer(", msl_options.indirect_params_buffer_index, ")]]");
+		}
+        else if (stage_out_var_id)
```
Comment (cdavis5e) [inline spirv_msl.cpp]: Formatting. This project uses tabs for indentation. (If you have `clang-format(1)` installed, you can use the `format_all.sh` script to automate this.)

Comment (mbarriault) [reply spirv_msl.cpp]: Corrected.

**Comment by mbarriault** [comment]
With KhronosGroup/MoltenVK#572 merged, is this clear to merge in now? MoltenVK tessellation is still broken without this commit.

**Comment by cdavis5e** [comment]
I have no objections. @HansKristian-Work?

**Comment by HansKristian-Work** [comment]
Then I have none either. Merged. Thanks!

**Pull request merged**